File: Tests/ClaudeUsageTests/UsageHistoryTests.swift
import XCTest
@testable import ClaudeUsage

final class UsageHistoryTests: XCTestCase {
    func testClaudeLocalTokensUseTodayAndDeduplicateMessages() async throws {
        let root = URL(fileURLWithPath: "/private/tmp/ClaudeUsage-local-token-fixture", isDirectory: true)
        defer { try? FileManager.default.removeItem(at: root) }
        try FileManager.default.createDirectory(at: root, withIntermediateDirectories: true)
        let now = Date()
        let today = ISO8601DateFormatter().string(from: now.addingTimeInterval(-60))
        let yesterday = ISO8601DateFormatter().string(from: now.addingTimeInterval(-25 * 60 * 60))
        let current = """
        {"requestId":"request-1","timestamp":"\(today)","message":{"id":"message-1","usage":{"input_tokens":10,"output_tokens":20,"cache_creation_input_tokens":30,"cache_read_input_tokens":40}}}
        """
        let old = """
        {"requestId":"request-2","timestamp":"\(yesterday)","message":{"id":"message-2","usage":{"input_tokens":999}}}
        """
        let fixture = [current, current, old].joined(separator: "\n")
        try Data(fixture.utf8).write(to: root.appendingPathComponent("session.jsonl"))

        let fetched = await ClaudeLocalTokenUsageService.fetch(now: now, rootURL: root)
        let usage = try XCTUnwrap(fetched)
        XCTAssertEqual(usage.todayTokens, 100)
    }

    @MainActor
    func testTrendUsesRecentSegmentAndTokenDelta() {
        let url = URL(fileURLWithPath: "/private/tmp/ClaudeUsage-history-test.json")
        defer { try? FileManager.default.removeItem(at: url) }
        let store = UsageHistoryStore(fileURL: url, minimumSampleInterval: 0)
        let now = Date(timeIntervalSince1970: 1_800_000_000)

        store.record(snapshot(pressure: 20, tokens: 1_000), at: now.addingTimeInterval(-3_600), force: true)
        store.record(snapshot(pressure: 28, tokens: 3_500), at: now, force: true)

        let trend = store.trend(now: now)
        XCTAssertEqual(trend.deltaPercent, 8)
        XCTAssertEqual(trend.percentPerHour, 8)
        XCTAssertEqual(trend.recentTokenDelta, 2_500)
        XCTAssertEqual(trend.points, [20, 28])
    }

    @MainActor
    func testResetStartsANewTrendSegmentAndRefreshesPet() {
        let url = URL(fileURLWithPath: "/private/tmp/ClaudeUsage-history-reset-test.json")
        defer { try? FileManager.default.removeItem(at: url) }
        let store = UsageHistoryStore(fileURL: url, minimumSampleInterval: 0)
        let now = Date(timeIntervalSince1970: 1_800_000_000)

        store.record(snapshot(pressure: 92, tokens: 10_000), at: now.addingTimeInterval(-1_800), force: true)
        store.record(snapshot(pressure: 8, tokens: 11_000), at: now.addingTimeInterval(-900), force: true)
        store.record(snapshot(pressure: 10, tokens: 12_000), at: now, force: true)

        let trend = store.trend(now: now)
        XCTAssertTrue(trend.resetDetected)
        XCTAssertEqual(trend.deltaPercent, 2)
        XCTAssertEqual(PetMood.resolve(snapshot: snapshot(pressure: 10, tokens: 12_000), trend: trend), .refreshed)
    }

    func testPetMoodUsesPressureAndBurnRate() {
        XCTAssertEqual(PetMood.resolve(snapshot: snapshot(pressure: 25, tokens: nil), trend: .empty), .calm)
        XCTAssertEqual(PetMood.resolve(snapshot: snapshot(pressure: 62, tokens: nil), trend: .empty), .focused)
        XCTAssertEqual(PetMood.resolve(snapshot: snapshot(pressure: 80, tokens: nil), trend: .empty), .sleepy)
        XCTAssertEqual(PetMood.resolve(snapshot: snapshot(pressure: 94, tokens: nil), trend: .empty), .tired)

        let fastTrend = UsageTrend(
            points: [20, 50],
            deltaPercent: 30,
            percentPerHour: 48,
            recentTokenDelta: nil,
            resetDetected: false
        )
        XCTAssertEqual(PetMood.resolve(snapshot: snapshot(pressure: 35, tokens: nil), trend: fastTrend), .tired)
    }

    func testMimoSensitivityChangesFocusedThreshold() {
        let current = snapshot(pressure: 42, tokens: nil)

        XCTAssertEqual(
            PetMood.resolve(snapshot: current, trend: .empty, sensitivity: .responsive),
            .focused
        )
        XCTAssertEqual(
            PetMood.resolve(snapshot: current, trend: .empty, sensitivity: .balanced),
            .calm
        )
        XCTAssertEqual(
            PetMood.resolve(snapshot: current, trend: .empty, sensitivity: .relaxed),
            .calm
        )
    }

    func testPressureUsesHighestIndividualModelCounter() {
        let snapshot = UsageHistorySnapshot(
            claudeFiveHour: 41,
            claudeWeekly: 23,
            claudeModelMaximum: 43,
            openAIFiveHour: 32,
            openAIWeekly: 8,
            openAIModelMaximum: nil,
            claudeTodayTokens: nil,
            openAITodayTokens: nil,
            claudeModelCounters: [
                UsageHistoryCounter(id: "seven_day_fable", label: "Claude Fable", utilization: 43)
            ]
        )

        XCTAssertEqual(snapshot.pressure, 43)
        XCTAssertEqual(snapshot.pressureSource?.id, "seven_day_fable")
        XCTAssertEqual(snapshot.pressureSource?.provider, .claude)
    }

    func testMimoAnimationModesUseAdaptiveCadence() {
        XCTAssertEqual(MimoAnimationMode.automatic.updateInterval(for: .calm), 1.4)
        XCTAssertEqual(MimoAnimationMode.automatic.updateInterval(for: .focused), 0.45)
        XCTAssertEqual(MimoAnimationMode.lively.updateInterval(for: .calm), 0.25)
        XCTAssertNil(MimoAnimationMode.still.updateInterval(for: .focused))
        XCTAssertEqual(MimoAnimationMode.automatic.transitionDuration(for: .calm), 0.22)
        XCTAssertEqual(MimoAnimationMode.lively.transitionDuration(for: .calm), 0.16)
    }

    func testHistoryDashboardSummaryRespectsProviderScope() {
        let first = UsageHistorySample(
            timestamp: Date(timeIntervalSince1970: 1_800_000_000),
            snapshot: UsageHistorySnapshot(
                claudeFiveHour: 80,
                claudeWeekly: 20,
                claudeModelMaximum: nil,
                openAIFiveHour: 35,
                openAIWeekly: 10,
                openAIModelMaximum: nil,
                claudeTodayTokens: nil,
                openAITodayTokens: nil
            )
        )
        let second = UsageHistorySample(
            timestamp: Date(timeIntervalSince1970: 1_800_000_300),
            snapshot: UsageHistorySnapshot(
                claudeFiveHour: 50,
                claudeWeekly: 20,
                claudeModelMaximum: nil,
                openAIFiveHour: 40,
                openAIWeekly: 10,
                openAIModelMaximum: nil,
                claudeTodayTokens: nil,
                openAITodayTokens: nil
            )
        )

        XCTAssertEqual(UsageHistoryDashboardView.pressure(in: second, scope: .claude), 50)
        XCTAssertEqual(UsageHistoryDashboardView.pressure(in: second, scope: .codex), 40)
        XCTAssertEqual(UsageHistoryDashboardView.detectedResetCount([first, second], scope: .claude), 1)
        XCTAssertEqual(UsageHistoryDashboardView.detectedResetCount([first, second], scope: .codex), 0)
    }

    @MainActor
    func testHistoryPersistsModelCountersAndLoadsLegacySamples() throws {
        let url = URL(fileURLWithPath: "/private/tmp/ClaudeUsage-history-schema-test.json")
        defer { try? FileManager.default.removeItem(at: url) }
        let store = UsageHistoryStore(fileURL: url, minimumSampleInterval: 0)
        let now = Date(timeIntervalSince1970: 1_800_000_000)
        let model = UsageHistoryCounter(
            id: "seven_day_fable",
            label: "Claude Fable",
            utilization: 47
        )

        store.record(
            UsageHistorySnapshot(
                claudeFiveHour: 30,
                claudeWeekly: 20,
                claudeModelMaximum: 47,
                openAIFiveHour: nil,
                openAIWeekly: nil,
                openAIModelMaximum: nil,
                claudeTodayTokens: nil,
                openAITodayTokens: nil,
                claudeModelCounters: [model]
            ),
            at: now,
            force: true
        )

        let reloaded = UsageHistoryStore(fileURL: url, minimumSampleInterval: 0)
        XCTAssertEqual(reloaded.samples.first?.claudeModelCounters, [model])

        let legacyJSON = """
        [{"timestamp":0,"claudeFiveHour":12,"claudeWeekly":8}]
        """
        try Data(legacyJSON.utf8).write(to: url, options: .atomic)
        let legacy = UsageHistoryStore(fileURL: url, minimumSampleInterval: 0)
        XCTAssertEqual(legacy.samples.first?.snapshot.pressure, 12)
        XCTAssertEqual(legacy.samples.first?.snapshot.claudeModelCounters, [])
    }

    @MainActor
    func testRecordingUsesFiveMinuteCadenceButCapturesReset() {
        let url = URL(fileURLWithPath: "/private/tmp/ClaudeUsage-history-cadence-test.json")
        defer { try? FileManager.default.removeItem(at: url) }
        let store = UsageHistoryStore(fileURL: url, minimumSampleInterval: 300)
        let now = Date(timeIntervalSince1970: 1_800_000_000)

        store.record(snapshot(pressure: 40, tokens: 1_000), at: now, force: true)
        store.record(snapshot(pressure: 48, tokens: 5_000), at: now.addingTimeInterval(60))
        XCTAssertEqual(store.samples.count, 1)

        store.record(snapshot(pressure: 15, tokens: 5_500), at: now.addingTimeInterval(120))
        XCTAssertEqual(store.samples.count, 2)
    }

    @MainActor
    func testRefreshedMoodExpiresThirtyMinutesAfterReset() {
        let url = URL(fileURLWithPath: "/private/tmp/ClaudeUsage-history-old-reset-test.json")
        defer { try? FileManager.default.removeItem(at: url) }
        let store = UsageHistoryStore(fileURL: url, minimumSampleInterval: 0)
        let now = Date(timeIntervalSince1970: 1_800_000_000)

        store.record(snapshot(pressure: 90, tokens: nil), at: now.addingTimeInterval(-2_400), force: true)
        store.record(snapshot(pressure: 10, tokens: nil), at: now.addingTimeInterval(-2_100), force: true)
        store.record(snapshot(pressure: 12, tokens: nil), at: now, force: true)

        XCTAssertFalse(store.trend(now: now).resetDetected)
    }

    private func snapshot(pressure: Double, tokens: Int64?) -> UsageHistorySnapshot {
        UsageHistorySnapshot(
            claudeFiveHour: pressure,
            claudeWeekly: nil,
            claudeModelMaximum: nil,
            openAIFiveHour: nil,
            openAIWeekly: nil,
            openAIModelMaximum: nil,
            claudeTodayTokens: tokens,
            openAITodayTokens: nil
        )
    }
}

File: windows/tests/ClaudeUsage.Core.Tests/UsageHistoryTests.cs
using System.Text.Json;
using ClaudeUsage.Core.History;
using ClaudeUsage.Core.Models;

namespace ClaudeUsage.Core.Tests;

public sealed class UsageHistoryTests
{
    [Fact]
    public void PressureUsesHighestIndividualCounterAndStableTieBreak()
    {
        var snapshot = new UsageHistorySnapshot(
            claudeFiveHour: 41,
            claudeWeekly: 23,
            claudeModelMaximum: 99,
            openAIFiveHour: 43,
            claudeModelCounters:
            [
                new UsageHistoryCounter("seven_day_zeta", "Claude Zeta", 43),
                new UsageHistoryCounter("seven_day_fable", "Claude Fable", 43),
            ]);

        Assert.Equal(43, snapshot.Pressure);
        Assert.Equal("codex-five-hour", snapshot.PressureSource?.Id);
        Assert.Equal(UsageProvider.Codex, snapshot.PressureSource?.Provider);
        Assert.Equal(43, snapshot.PressureFor(UsageProvider.Claude));
        Assert.DoesNotContain(snapshot.PressureSources, source => source.Id == "claude-model-maximum");
    }

    [Fact]
    public void TodayTokensSumAvailableProviders()
    {
        Assert.Null(new UsageHistorySnapshot().TodayTokens);
        Assert.Equal(120, new UsageHistorySnapshot(claudeTodayTokens: 120).TodayTokens);
        Assert.Equal(
            320,
            new UsageHistorySnapshot(claudeTodayTokens: 120, openAITodayTokens: 200).TodayTokens);
    }

    [Fact]
    public void TrendUsesRecentSegmentAndTokenDelta()
    {
        using var fixture = new HistoryFixture();
        var store = fixture.CreateStore(minimumSampleInterval: TimeSpan.Zero);
        var now = DateTimeOffset.FromUnixTimeSeconds(1_800_000_000);

        store.Record(Snapshot(20, 1_000), now.AddHours(-1), force: true);
        store.Record(Snapshot(28, 3_500), now, force: true);

        var trend = store.Trend(now, calendarTimeZone: TimeZoneInfo.Utc);

        Assert.Equal(8, trend.DeltaPercent);
        Assert.Equal(8, trend.PercentPerHour);
        Assert.Equal(2_500, trend.RecentTokenDelta);
        Assert.Equal([20d, 28d], trend.Points);
    }

    [Fact]
    public void ResetStartsNewTrendSegmentAndRefreshesCompanion()
    {
        using var fixture = new HistoryFixture();
        var store = fixture.CreateStore(minimumSampleInterval: TimeSpan.Zero);
        var now = DateTimeOffset.FromUnixTimeSeconds(1_800_000_000);

        store.Record(Snapshot(92, 10_000), now.AddMinutes(-30), force: true);
        store.Record(Snapshot(8, 11_000), now.AddMinutes(-15), force: true);
        store.Record(Snapshot(10, 12_000), now, force: true);

        var trend = store.Trend(now, calendarTimeZone: TimeZoneInfo.Utc);

        Assert.True(trend.ResetDetected);
        Assert.Equal(2, trend.DeltaPercent);
        Assert.Equal(PetMood.Refreshed, PetMoodResolver.Resolve(Snapshot(10, 12_000), trend));
    }

    [Theory]
    [InlineData(25, PetMood.Calm)]
    [InlineData(62, PetMood.Focused)]
    [InlineData(80, PetMood.Sleepy)]
    [InlineData(94, PetMood.Tired)]
    public void BalancedMoodUsesPressure(double pressure, PetMood expected)
    {
        Assert.Equal(expected, PetMoodResolver.Resolve(Snapshot(pressure, null)));
    }

    [Fact]
    public void MoodUsesPaceAndSensitivityThresholds()
    {
        var rapid = new UsageTrend([20, 50], 30, 48, null, false);
        Assert.Equal(PetMood.Tired, PetMoodResolver.Resolve(Snapshot(35, null), rapid));

        var current = Snapshot(42, null);
        Assert.Equal(
            PetMood.Focused,
            PetMoodResolver.Resolve(current, UsageTrend.Empty, MimoSensitivity.Responsive));
        Assert.Equal(
            PetMood.Calm,
            PetMoodResolver.Resolve(current, UsageTrend.Empty, MimoSensitivity.Balanced));
        Assert.Equal(
            PetMood.Calm,
            PetMoodResolver.Resolve(current, UsageTrend.Empty, MimoSensitivity.Relaxed));
    }

    [Fact]
    public void RecordingUsesFiveMinuteCadenceButCapturesReset()
    {
        using var fixture = new HistoryFixture();
        var store = fixture.CreateStore(minimumSampleInterval: TimeSpan.FromMinutes(5));
        var now = DateTimeOffset.FromUnixTimeSeconds(1_800_000_000);

        Assert.True(store.Record(Snapshot(40, 1_000), now, force: true));
        Assert.False(store.Record(Snapshot(48, 5_000), now.AddMinutes(1)));
        Assert.Single(store.Samples);

        Assert.True(store.Record(Snapshot(15, 5_500), now.AddMinutes(2)));
        Assert.Equal(2, store.Samples.Count);
    }

    [Fact]
    public void RefreshedMoodExpiresThirtyMinutesAfterReset()
    {
        using var fixture = new HistoryFixture();
        var store = fixture.CreateStore(minimumSampleInterval: TimeSpan.Zero);
        var now = DateTimeOffset.FromUnixTimeSeconds(1_800_000_000);

        store.Record(Snapshot(90, null), now.AddMinutes(-40), force: true);
        store.Record(Snapshot(10, null), now.AddMinutes(-35), force: true);
        store.Record(Snapshot(12, null), now, force: true);

        Assert.False(store.Trend(now).ResetDetected);
    }

    [Fact]
    public void RetentionIsBoundedByAgeAndCount()
    {
        using var fixture = new HistoryFixture();
        var store = new UsageHistoryStore(
            fixture.FilePath,
            new UsageHistoryOptions
            {
                MinimumSampleInterval = TimeSpan.Zero,
                RetentionInterval = TimeSpan.FromDays(14),
                MaximumSamples = 3,
            });
        var now = DateTimeOffset.FromUnixTimeSeconds(1_800_000_000);

        store.Record(Snapshot(10, null), now.AddDays(-20), force: true);
        store.Record(Snapshot(20, null), now.AddDays(-2), force: true);
        store.Record(Snapshot(30, null), now.AddDays(-1), force: true);
        store.Record(Snapshot(40, null), now.AddHours(-1), force: true);
        store.Record(Snapshot(50, null), now, force: true);

        Assert.Equal(3, store.Samples.Count);
        Assert.Equal([30d, 40d, 50d], store.Samples.Select(sample => sample.Snapshot.Pressure!.Value));
        Assert.All(store.Samples, sample => Assert.True(sample.Timestamp >= now.AddDays(-14)));
    }

    [Fact]
    public void HistoryPersistsModelCountersAndLoadsLegacySamples()
    {
        using var fixture = new HistoryFixture();
        var now = DateTimeOffset.FromUnixTimeSeconds(1_800_000_000);
        var model = new UsageHistoryCounter("seven_day_fable", "Claude Fable", 47);
        var store = fixture.CreateStore(minimumSampleInterval: TimeSpan.Zero);

        store.Record(
            new UsageHistorySnapshot(
                claudeFiveHour: 30,
                claudeWeekly: 20,
                claudeModelMaximum: 47,
                claudeModelCounters: [model]),
            now,
            force: true);

        var reloaded = fixture.CreateStore(minimumSampleInterval: TimeSpan.Zero);
        Assert.Equal(model, Assert.Single(Assert.Single(reloaded.Samples).ClaudeModelCounters!));

        File.WriteAllText(
            fixture.FilePath,
            "[{\"timestamp\":\"2027-01-15T08:00:00+00:00\",\"claudeFiveHour\":12,\"claudeWeekly\":8}]");
        var legacy = fixture.CreateStore(minimumSampleInterval: TimeSpan.Zero);
        var legacySample = Assert.Single(legacy.Samples);
        Assert.Equal(12, legacySample.Snapshot.Pressure);
        Assert.Empty(legacySample.Snapshot.ClaudeModelCounters);
    }

    [Fact]
    public void StoreSkipsEmptySnapshotAndClearDeletesOnlyHistoryFile()
    {
        using var fixture = new HistoryFixture();
        var store = fixture.CreateStore(minimumSampleInterval: TimeSpan.Zero);

        Assert.False(store.Record(new UsageHistorySnapshot(), force: true));
        Assert.False(File.Exists(fixture.FilePath));

        store.Record(Snapshot(10, null), force: true);
        Assert.True(File.Exists(fixture.FilePath));
        store.Clear();

        Assert.Empty(store.Samples);
        Assert.False(File.Exists(fixture.FilePath));
    }

    [Fact]
    public void ReopenImmediatelyPrunesExpiredAndOversizedHistory()
    {
        using var fixture = new HistoryFixture();
        var now = DateTimeOffset.Now;
        var samples = new[]
        {
            new UsageHistorySample(now.AddDays(-30), Snapshot(5, null)),
            new UsageHistorySample(now.AddDays(-4), Snapshot(10, null)),
            new UsageHistorySample(now.AddDays(-3), Snapshot(20, null)),
            new UsageHistorySample(now.AddDays(-2), Snapshot(30, null)),
            new UsageHistorySample(now.AddDays(-1), Snapshot(40, null)),
        };
        File.WriteAllText(fixture.FilePath, JsonSerializer.Serialize(samples));

        var options = new UsageHistoryOptions
        {
            MinimumSampleInterval = TimeSpan.Zero,
            RetentionInterval = TimeSpan.FromDays(14),
            MaximumSamples = 3,
        };
        var store = new UsageHistoryStore(fixture.FilePath, options);

        Assert.Equal([20d, 30d, 40d], store.Samples.Select(sample => sample.Snapshot.Pressure!.Value));
        Assert.DoesNotContain(store.Samples, sample => sample.Timestamp < now.AddDays(-14));

        var reloaded = new UsageHistoryStore(fixture.FilePath, options);
        Assert.Equal(3, reloaded.Samples.Count);
    }

    [Fact]
    public void ClearRemovesCrashTemporaryHistoryFiles()
    {
        using var fixture = new HistoryFixture();
        var store = fixture.CreateStore(minimumSampleInterval: TimeSpan.Zero);
        store.Record(Snapshot(10, null), force: true);
        var temporaryPath = Path.Combine(
            Path.GetDirectoryName(fixture.FilePath)!,
            $".{Path.GetFileName(fixture.FilePath)}.deadbeef.tmp");
        File.WriteAllText(temporaryPath, "sensitive-history-remnant");

        store.Clear();

        Assert.False(File.Exists(fixture.FilePath));
        Assert.False(File.Exists(temporaryPath));
    }

    [Fact]
    public void ClearReportsFailureWhenHistoryPathCannotBeDeleted()
    {
        using var fixture = new HistoryFixture();
        Directory.CreateDirectory(fixture.FilePath);
        var store = fixture.CreateStore(minimumSampleInterval: TimeSpan.Zero);

        var exception = Assert.Throws<IOException>(() => store.Clear());

        Assert.Contains("could not be deleted completely", exception.Message);
        Assert.NotNull(store.LastPersistenceError);
        Assert.True(Directory.Exists(fixture.FilePath));
    }

    [Fact]
    public void LoadIgnoresNullSamplesAndNullModelCounters()
    {
        using var fixture = new HistoryFixture();
        var timestamp = DateTimeOffset.Now;
        var timestampJson = JsonSerializer.Serialize(timestamp);
        File.WriteAllText(
            fixture.FilePath,
            $$"""
            [
              null,
              {
                "timestamp": {{timestampJson}},
                "claudeFiveHour": 21,
                "claudeModelCounters": [
                  null,
                  { "id": "seven_day_fable", "label": "Claude Fable", "utilization": 42 }
                ]
              }
            ]
            """);

        var store = fixture.CreateStore(minimumSampleInterval: TimeSpan.Zero);
        var sample = Assert.Single(store.Samples);

        Assert.Equal(42, sample.Snapshot.Pressure);
        Assert.Equal("seven_day_fable", Assert.Single(sample.Snapshot.ClaudeModelCounters).Id);
    }

    [Fact]
    public void CompanionCatalogAndAnimationCadenceAreStable()
    {
        Assert.Equal(
            [
                CompanionKind.Mimo,
                CompanionKind.Lumi,
                CompanionKind.Kumo,
                CompanionKind.Dot,
                CompanionKind.Navi,
                CompanionKind.Bori,
                CompanionKind.Muru,
                CompanionKind.Tori,
                CompanionKind.Pico,
            ],
            Enum.GetValues<CompanionKind>());
        Assert.Equal(
            ["mimo", "lumi", "kumo", "dot", "navi", "bori", "muru", "tori", "pico"],
            Enum.GetValues<CompanionKind>().Select(kind => kind.Id()));

        var encoded = JsonSerializer.Serialize(Enum.GetValues<CompanionKind>());
        Assert.Equal("[\"mimo\",\"lumi\",\"kumo\",\"dot\",\"navi\",\"bori\",\"muru\",\"tori\",\"pico\"]", encoded);
        Assert.Equal(
            Enum.GetValues<CompanionKind>(),
            JsonSerializer.Deserialize<CompanionKind[]>(encoded));

        Assert.Equal(1.4, MimoAnimationMode.Automatic.UpdateInterval(PetMood.Calm)?.TotalSeconds);
        Assert.Equal(0.45, MimoAnimationMode.Automatic.UpdateInterval(PetMood.Focused)?.TotalSeconds);
        Assert.Equal(0.25, MimoAnimationMode.Lively.UpdateInterval(PetMood.Calm)?.TotalSeconds);
        Assert.Null(MimoAnimationMode.Still.UpdateInterval(PetMood.Focused));
        Assert.Equal(0.22, MimoAnimationMode.Automatic.TransitionDuration(PetMood.Calm).TotalSeconds);
        Assert.Equal(0.16, MimoAnimationMode.Lively.TransitionDuration(PetMood.Calm).TotalSeconds);

        var first = CompanionPoseResolver.Resolve(CompanionKind.Mimo, PetMood.Focused, 100);
        Assert.Equal(first, CompanionPoseResolver.Resolve(CompanionKind.Mimo, PetMood.Focused, 100));
        Assert.NotEqual(first, CompanionPoseResolver.Resolve(CompanionKind.Mimo, PetMood.Focused, 101));
        var reduced = CompanionPoseResolver.Resolve(
            CompanionKind.Mimo,
            PetMood.Refreshed,
            long.MaxValue,
            reducedMotion: true);
        Assert.True(reduced.ShowCelebration);
        Assert.Equal(16, reduced.LeftPartAngle);
        Assert.Equal(-132, reduced.RightPartAngle);
        Assert.Equal(-4, reduced.LeftLegAngle);
        Assert.Equal(4, reduced.RightLegAngle);
    }

    [Fact]
    public void ReducedMotionUsesEachMacCharactersTimeZeroPose()
    {
        var cases = new[]
        {
            new PoseCase(CompanionKind.Mimo, PetMood.Tired, 34, -34, 0, 12, -12, 0, 1),
            new PoseCase(CompanionKind.Lumi, PetMood.Tired, 0, 0, 18, 0, 0, 0, 0.52),
            new PoseCase(CompanionKind.Bori, PetMood.Focused, 0, 0, 0, 0, 0, -48, 1),
            new PoseCase(CompanionKind.Bori, PetMood.Tired, 0, 0, 9, 0, 0, -18, 1),
            new PoseCase(CompanionKind.Muru, PetMood.Sleepy, 0, 0, 8, 0, 0, 0, 1),
            new PoseCase(CompanionKind.Tori, PetMood.Refreshed, -22, 22, 0, 0, 0, 0, 1),
            new PoseCase(CompanionKind.Pico, PetMood.Tired, -18, 18, 0, 0, 0, 42, 1),
            new PoseCase(CompanionKind.Dot, PetMood.Focused, 0, 0, 0, 0, 0, 0, 1),
            new PoseCase(CompanionKind.Navi, PetMood.Calm, 0, 0, 0, 0, 0, 0, 1),
        };

        foreach (var item in cases)
        {
            var pose = CompanionPoseResolver.Resolve(item.Kind, item.Mood, long.MaxValue, reducedMotion: true);
            Assert.Equal(item.LeftPartAngle, pose.LeftPartAngle);
            Assert.Equal(item.RightPartAngle, pose.RightPartAngle);
            Assert.Equal(item.HeadAngle, pose.HeadAngle);
            Assert.Equal(item.LeftLegAngle, pose.LeftLegAngle);
            Assert.Equal(item.RightLegAngle, pose.RightLegAngle);
            Assert.Equal(item.TailAngle, pose.TailAngle);
            Assert.Equal(0, pose.VerticalOffset);
            Assert.Equal(item.EffectOpacity, pose.EffectOpacity);
        }
    }

    private static UsageHistorySnapshot Snapshot(double pressure, long? tokens) => new(
        claudeFiveHour: pressure,
        claudeTodayTokens: tokens);

    private sealed record PoseCase(
        CompanionKind Kind,
        PetMood Mood,
        double LeftPartAngle,
        double RightPartAngle,
        double HeadAngle,
        double LeftLegAngle,
        double RightLegAngle,
        double TailAngle,
        double EffectOpacity);

    private sealed class HistoryFixture : IDisposable
    {
        private readonly string _directory = Path.Combine(
            Path.GetTempPath(),
            $"ClaudeUsage-history-{Guid.NewGuid():N}");

        public HistoryFixture()
        {
            Directory.CreateDirectory(_directory);
        }

        public string FilePath => Path.Combine(_directory, "usage-history.json");

        public UsageHistoryStore CreateStore(TimeSpan minimumSampleInterval) => new(
            FilePath,
            new UsageHistoryOptions { MinimumSampleInterval = minimumSampleInterval });

        public void Dispose()
        {
            try
            {
                Directory.Delete(_directory, recursive: true);
            }
            catch (IOException)
            {
            }
            catch (UnauthorizedAccessException)
            {
            }
        }
    }
}
